[WebGPU] present() needs to wait for all command buffers to be scheduled

litherum opened the pull request:
#### 442fabee7226138865b76fc2efded6e4aa4e4a90
<pre>
[WebGPU] present() needs to wait for all command buffers to be scheduled
<a href="https://bugs.webkit.org/show_bug.cgi?id=250993">https://bugs.webkit.org/show_bug.cgi?id=250993</a>
rdar://104538554

Reviewed by Dean Jackson.

We already had a callback set up in prepareForDisplay(). All we need to do is plumb that through to
WebGPU::Queue, and use the same kind of infrastructure that we already have for onSubmittedWorkDone(),
but for onSubmittedWorkScheduled().

* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUCompositorIntegrationImpl.cpp:
(PAL::WebGPU::CompositorIntegrationImpl::prepareForDisplay):
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUCompositorIntegrationImpl.h:
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUImpl.cpp:
(PAL::WebGPU::WTF::Function&lt;void):
(PAL::WebGPU::GPUImpl::createPresentationContext):
* Source/WebGPU/WebGPU/PresentationContextIOSurface.h:
* Source/WebGPU/WebGPU/PresentationContextIOSurface.mm:
(WebGPU::PresentationContextIOSurface::create):
(WebGPU::PresentationContextIOSurface::onSubmittedWorkScheduled):
(WebGPU::PresentationContextIOSurface::configure):
(WebGPU::PresentationContextIOSurface::unconfigure):
* Source/WebGPU/WebGPU/Queue.h:
(WebGPU::Queue::isSchedulingIdle const):
* Source/WebGPU/WebGPU/Queue.mm:
(WebGPU::Queue::onSubmittedWorkScheduled):
(WebGPU::Queue::commitMTLCommandBuffer):
* Source/WebGPU/WebGPU/WebGPUExt.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteCompositorIntegration.messages.in:

Canonical link: <a href="https://commits.webkit.org/259912@main">https://commits.webkit.org/259912@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f5ea971b53deb0fa1aa7c98bc58936916aa00804

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106252 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15306 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/39090 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115443 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/175541 "Failed to checkout and rebase branch from PR 8961") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/110160 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16748 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6522 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98466 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/115125 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/112014 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12733 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95710 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/40285 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94607 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27353 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/81974 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8547 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28705 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/9054 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/5282 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14669 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48251 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6848 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10590 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebGPU/WebGPU/Queue.h:
```diff
@@ -96,6 +98,48 @@ class Queue : public WGPUQueueImpl, public ThreadSafeRefCounted<Queue> {
+        {
+        }
+
+        NSCondition *m_condition { nil };
```
What is the resolution of NSCondition and can this be implemented without a lock?

litherum replied:
What does "resolution" mean here? Are you asking which state it's guarding?

mwyrzykowski replied:
Oh sorry I meant how quickly does `wait` return when `signal` is called from the other thread? Maybe that is an implementation detail outside the scope of this review.

mwyrzykowski commented on Source/WebGPU/WebGPU/Queue.mm:
```diff
@@ -124,20 +126,32 @@
 {
     ASSERT(commandBuffer.commandQueue == m_commandQueue);
     [commandBuffer addCompletedHandler:[protectedThis = Ref { *this }](id<MTLCommandBuffer>) {
+        // We're not on the main thread right now, so immediately bounce on over to the main thread.
```
I don't think this comment does what is says it does. `scheduleWork` does not run on the main thread right?

litherum replied:
The semantics of scheduleWork are described in a comment in WebGPUExt.h. It's supposed to schedule work to occur on the main thread, regardless of which thread it's called from.

mwyrzykowski replied:
In the GPU Process scheduleWork runs on the WebGPU work queue:

from RemoteGPU.cpp:89
```
    auto backing = PAL::WebGPU::create([protectedThis = Ref { *this }](PAL::WebGPU::WorkItem&& workItem) {
        protectedThis->workQueue().dispatch(WTFMove(workItem));
    });
```

which can be seen in the backtrace:
```
WebGPU work queue (4)#0	0x0000000104bffb18 in WebGPU::Queue::commitMTLCommandBuffer(id<MTLCommandBuffer>)::$_1::operator()(objc_object<MTLCommandBuffer>) const::'lambda'()::operator()() const at /Volumes/Data/OpenSource/Source/WebGPU/WebGPU/Queue.mm:129
#1	0x0000000104bff8b8 in WTF::Detail::CallableWrapper<WebGPU::Queue::commitMTLCommandBuffer(id<MTLCommandBuffer>)::$_1::operator()(objc_object<MTLCommandBuffer>) const::'lambda'(), void>::call() at /Volumes/Data/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/Function.h:53
```

litherum replied:
Ah, yes, I was thinking of the work queue as the main thread; you're right that those are different. I should update the comment to say something like:

```
// We're on a dispatch queue that's internal to Metal right now, so let's immediately bounce on over to the correct work queue to avoid concurrent accesses.
```

mwyrzykowski replied:
If we are just blocking the gpu work queue instead of the main thread then that seems like less of an issue. However I think it would be cleaner without locks. What happens to the work queue's thread priority when it is blocked waiting for the signal for instance? These issues we would not have to worry about if we are not using locks.

mwyrzykowski reviewed (commented):
I think it would be fairly straight forwards to implement this without NSCondition and locking.

Metal already gives us callbacks for when the MTLCommandBuffer instances are scheduled via addScheduledHandler which is used in this patch.

SwapChainImpl::prepareForDisplay is already async. It could be changed to:

```
void SwapChainImpl::prepareForDisplay(CompletionHandler<void(WTF::MachSendRight&&)>&& completionHandler)
{
  wgpuSwapChainPresentWithCompletion(m_backing, ^(ioSurface) {
      completionHandler(MachSendRight::adopt(IOSurfaceCreateMachPort(ioSurface)));
  });
}
```

Then `PresentationContextIOSurface::present()` becomes:
```
void PresentationContextIOSurface::present(completionHandler)
{
   auto block = ^{
          nextDrawable();
          completionHandler(m_displayBuffer);
   };
    if (!m_device) {
      block();
      return;
     }

     m_device->getQueue().runCompletionHandlerAfterAllCommandBuffersScheduled(block);
}
```

maybe ensuring `block` runs on the GPU work queue.

mwyrzykowski reviewed (changes requested):
I think we should try to implement this without NSCondition and locks, it seems to be unnecessary.

litherum committed:
[WebGPU] present() needs to wait for all command buffers to be scheduled
https://bugs.webkit.org/show_bug.cgi?id=250993
rdar://104538554

Reviewed by Dean Jackson.

We already had a callback set up in prepareForDisplay(). All we need to do is plumb that through to
WebGPU::Queue, and use the same kind of infrastructure that we already have for onSubmittedWorkDone(),
but for onSubmittedWorkScheduled().

* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUCompositorIntegrationImpl.cpp:
(PAL::WebGPU::CompositorIntegrationImpl::prepareForDisplay):
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUCompositorIntegrationImpl.h:
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUImpl.cpp:
(PAL::WebGPU::WTF::Function<void):
(PAL::WebGPU::GPUImpl::createPresentationContext):
* Source/WebGPU/WebGPU/PresentationContextIOSurface.h:
* Source/WebGPU/WebGPU/PresentationContextIOSurface.mm:
(WebGPU::PresentationContextIOSurface::create):
(WebGPU::PresentationContextIOSurface::onSubmittedWorkScheduled):
(WebGPU::PresentationContextIOSurface::configure):
(WebGPU::PresentationContextIOSurface::unconfigure):
* Source/WebGPU/WebGPU/Queue.h:
(WebGPU::Queue::isSchedulingIdle const):
* Source/WebGPU/WebGPU/Queue.mm:
(WebGPU::Queue::onSubmittedWorkScheduled):
(WebGPU::Queue::commitMTLCommandBuffer):
* Source/WebGPU/WebGPU/WebGPUExt.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteCompositorIntegration.messages.in:

Canonical link: https://commits.webkit.org/259912@main
